[BE]할 일 조회 API 구현

2heekeon opened the pull request.

2heekeon committed:
feat[#35]: 할 일 수정 API 구현

- requestParam으로 id 받기
- RuntimeException 상속받은 ToDoNotFoundException 생성

2heekeon committed:
feat[#35]: 할 일 수정 API 구현

- requestParam으로 id 받기
- RuntimeException 상속받은 ToDoNotFoundException 생성

2heekeon committed:
feat[#36]: 할 일 삭제 API 구현

- requestParam으로 id 받기
- RuntimeException 상속받은 ToDoNotFoundException 생성

2heekeon committed:
feat[#40]: 할 일 단건 조회, 날짜별 조회 API 구현

2heekeon committed:
feat[#40]: 컨트롤러 RequestParam 기본값 설정
서비스에 핸들러 메소드 구현 및 로직 간단화

choitree commented on src/main/java/com/codeQuartette/myTime/service/impl/ToDoServiceImpl.java:
```diff
@@ -44,8 +39,39 @@ public void create(Long userId, ToDoDTO.Request toDoRequestDTO) {
+    }
+
+    //한건 조회
+    @Override
+    public ToDoDTO.Response getToDoById(Long id) {
+        return ToDoDTO.Response.of(handler(id));
+    }
```
리턴 타입을 ToDo로 바꾸고, 메소드명을 findToDo로 변경 -> 위에 해당 메소드가 존재하기 때문에 삭제하면 되는 메소드

service layer에서는 return type을 모두 Entity 또는 List<Entity>로 통일하고,
controller layer에서 Entity를 Response로 변경하는 작업 진행

2heekeon replied:
ToDoDTO.Response.of 메소드를 List<ToDo> find
리턴 타입을 List로 통일 완료, Entity를 Response로 변경 완료

malaheaven commented on src/main/java/com/codeQuartette/myTime/controller/ToDoController.java:
```diff
@@ -15,7 +15,39 @@ public class ToDoController {
+    @PatchMapping("/todo")
+    public void update(@RequestParam(name = "id", required = false) Long id,
+                       @RequestBody ToDoDTO.Request toDoRequestDTO) {
+        toDoService.update(id, toDoRequestDTO);
```
희건님은 update하고 수정된 객체를 return을 해주지 않는데,
저는 수정된 객체를 return 해주도록 개발했습니다.

저희 팀끼리 수정 API 마다 return 값을 어떻게 해줄건지 상의를 해볼 필요가 있어보입니다.

malaheaven commented on src/main/java/com/codeQuartette/myTime/controller/ToDoController.java:
```diff
@@ -15,7 +15,39 @@ public class ToDoController {
+    }
+
+    //한건 조회
+    @GetMapping("/todo")
+    public ToDoDTO.Response getToDoById(@RequestParam(name = "id", required = false) Long id) {
+        return toDoService.getToDoById(id);
+    }
+
+    //날짜별 조회
+    @GetMapping("/toDos")
+    public List<ToDoDTO.Response> getToDoByDate(@RequestParam(name = "userId", required = false) Long userId,
+                                                @RequestParam(name = "date", required = false) LocalDate date) {
+        return toDoService.getToDoByDate(userId, date);
```
List<ToDoDTO.Response> 로 합칠 수 있어보입니다.
한건 조회도 List<ToDoDTO.Response>를 사용해 return될 수 있도록 하면 좋을 거 같습니다.

그렇게 되면 GetMapping, DeleteMapping, PatchMapping, PostMapping  EndPoint 모두가 /todo로 동일하니
@ RequestMapping 어노테이션을 사용해 중복을 줄일 수 있어보입니다.

2heekeon replied:
한 건 조회와 날짜 별 조회 합치고 메소드명은 ResponseEntity<ToDoDTO.ResponseList>로 변경
@ RequestMapping 어노테이션 적용 완료

malaheaven commented on src/main/java/com/codeQuartette/myTime/controller/dto/ToDoDTO.java:
```diff
@@ -1,26 +1,53 @@
+
+        private Boolean isBlind;
 
+        public static ToDoDTO.Response of(ToDo toDo) {
```
현재 사용한 메서드는 정적팩토리 메서드라고 부릅니다.
한번 찾아보시면 좋을 것 같아 코멘트 남깁니다.

2heekeon replied:
감사합니다!

malaheaven commented on src/main/java/com/codeQuartette/myTime/domain/ToDo.java:
```diff
@@ -31,7 +33,6 @@ public class ToDo {
 
-
     public static ToDo create(ToDoDTO.Request toDoRequestDTO) {
         return com.codeQuartette.myTime.domain.ToDo.builder()
```
import에 올라가져야할 부분이 메소드처럼 사용되었네요 수정 바랍니다.

2heekeon replied:
com.codeQuartette.myTime.domain  import 로 수정 완료

eNoLJ commented on src/main/java/com/codeQuartette/myTime/service/impl/ToDoServiceImpl.java:
```diff
@@ -44,8 +39,39 @@ public void create(Long userId, ToDoDTO.Request toDoRequestDTO) {
+    @Override
+    public List<ToDoDTO.Response> getToDoByDate(Long userId, LocalDate date) {
+        MyDate myDate = myDateService.find(date).stream()
+                .filter(urDate -> urDate.getUser().getId().equals(userId))
```
filter 안에 변수명을 urDate처럼 줄임말보다 좀 길어도 되니 명시적으로 적어주는게 좋아보입니다.
저는 이런 경우 targetDate나 currentDate를 선호하는 편인데 편하실대로 수정해주세요.

그리고
`urDate.getUser().getId().equals(userId))` 부분을 객체별로 역할을 나눠 부여해주세요.
위에서는
`urDate.matchUser(userId)` 정도로 줄일 수 있어보입니다.
그리고 User 엔티티 클래스에서
`public boolean matchId(userId)` 메소드를 만들어 id가 일치하는지 직접 User에게 물어볼 수 있도록 해주세요.
그리고 urDate 클래스에서 해당 메소드를 사용해주면 객체 별로 역할을 나눌 수 있을 것 같아요.
메소드명같은 경우는 제가 즐겨 사용하는 방식으로 예를 들었는데 직접 작성하실 때 좀 더 어울리는 메소드명이 있다면 사용하시면 되겠습니다.

2heekeon committed:
refactor[#63]:조회 로직 리팩토링

변수명 변경 urDate -> targetDate
filter 부분 객체별 역할 분배

2heekeon committed:
refactor[#63]:조회 클래스명 변경

클래스, 메소드명 변경 ToDoDTO.Response -> List<ToDo>
 getToDoById -> find

2heekeon committed:
refactor[#63]:조회 컨트롤러 병합

클래스명 변경 getToDoById 와 getToDoByDate -> ResponseEntity<ToDoDTO.ResponseList>
로직 수정

2heekeon committed:
refactor[#63]:import에 올라가져야할 부분이 메소드로 수정

불필요 메소드 삭제

2heekeon committed:
refactor[#63]:RequestMapping 어노테이션 사용

@RequestMapping("/todo")로 통합

2heekeon committed:
refactor[#63]: 불필요 메소드 수정

2heekeon committed:
Merge branch 'dev' into be/feature/todo
